libs/dmlf/src/basic_vm_engine.cpp

```diff
@@ -389,117 +196,267 @@ bool BasicVmEngine::HasState(std::string const &name) const
   return states_.find(name) != states_.end();
 }
 
-bool BasicVmEngine::Convertable(LedgerVariant const &ledgerVariant, TypeId const &typeId) const
+BasicVmEngine::Error BasicVmEngine::PrepInput(vm::ParameterPack &result, Params const &params,
+                                              VM &vm, Executable *exec,
+                                              Executable::Function const *func,
+                                              std::string const &         runName)
 {
-  switch (typeId)
-  {
-  case fetch::vm::TypeIds::Bool:
+  const std::string errorPrefix = "Error( " + runName + "): ";
+
+  auto const numParameters = static_cast<std::size_t>(func->num_parameters);
+  if (numParameters != params.size())
   {
-    return ledgerVariant.IsBoolean();
+    return Error(Error::Stage::ENGINE, Error::Code::RUNTIME_ERROR,
+                 errorPrefix + "Wrong number of parameters expected " +
+                     std::to_string(numParameters) + "; received " + std::to_string(params.size()));
   }
-  case fetch::vm::TypeIds::Int8:
-  case fetch::vm::TypeIds::UInt8:
-  case fetch::vm::TypeIds::Int16:
-  case fetch::vm::TypeIds::UInt16:
-  case fetch::vm::TypeIds::Int32:
-  case fetch::vm::TypeIds::UInt32:
-  case fetch::vm::TypeIds::Int64:
+
+  // Serialize LedgerVariant to MsgPack...
+  serializers::MsgPackSerializer serializer;
+  try
   {
-    return ledgerVariant.IsInteger();
+    for (auto const &p : params)
+    {
+      serializer << p;
+    }
   }
-  case fetch::vm::TypeIds::Float32:
-  case fetch::vm::TypeIds::Float64:
+  catch (std::exception &ex)
   {
-    return ledgerVariant.IsFloatingPoint();
+    return Error(
+        Error::Stage::ENGINE, Error::Code::SERIALIZATION_ERROR,
+        errorPrefix + "Serializing input before running: Threw error " + std::string(ex.what()));
```

Comment:
Is `std::string(` needed here?
Btw, `errorPrefix` could be declared non-const above and then `move`d here, and on other sites.

Reply:
The std::string is needed in some of the others so I kept it as is for consistency. 
I prefer to declare it const as I think it declares the intent in a better way.